### PR TITLE
Refactor GetNextResourceIndex to make it work properly with transaction

### PR DESCRIPTION
**Pull request by wxiaoguang**
Related:
* #21362

This PR uses a general and stable method to generate resource index (eg: Issue Index, PR Index)

If the code looks good, I can add more tests

ps: please skip the diff, only have a look at the new code. It's entirely re-written.

### CODE REVIEW

**Review by Gusted — APPROVED** [review]
Code-wise looks good. I'll trust you on the write-lock SQL code, no expertise on that.

**Comment by wxiaoguang** [comment]
Added tests.

Ready for review & merge.

File: models/db/index.go [hunk 1]
```diff
@@ -59,53 +29,89 @@ const (
...
+	res, err := e.Exec(fmt.Sprintf("UPDATE %s SET max_index=? WHERE group_id=? AND max_index<?", tableName), maxIndex, groupID, maxIndex)
+	if err != nil {
+		return err
+	}
+	affected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if affected == 0 {
+		// if nothing is updated, the record might not exist, try to insert and update it
```
Comment (lunny) [inline models/db/index.go]: Another possibility is maxIndex is equal or less than the value stored in the resource_index table.

Comment (wxiaoguang) [reply models/db/index.go]: Yes, the code also handles it correctly, see the tests.

Feel free to edit.

Comment (wxiaoguang) [inline models/db/index.go]:
```suggestion
		// if nothing is updated, the record might not exist or is larger, it's safe to try to insert and update it again
```

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Maybe we need to ensure `SyncMaxResourceIndex` or `GetNextResourceIndex` is in a transaction.

**Comment by wxiaoguang** [comment]
> Maybe we need to ensure `SyncMaxResourceIndex` or `GetNextResourceIndex` is in a transaction.

Agree. Although the current approach also works without transaction (it's designed to be so), I hope there is a mechanism to detect whether the context has an open transaction. If there is such mechanism, we can update the code to add the checks.

**Comment by lunny** [comment]
make L-G-T-M work

**Pull request merged**